src/whitenoise/storage/media_files.rs

```diff
@@ -0,0 +1,314 @@
+use crate::whitenoise::error::{Result, WhitenoiseError};
+use mdk_core::GroupId;
+use std::path::{Path, PathBuf};
+use tokio::fs;
+
+/// Filesystem storage for media files organized by MLS group
+///
+/// Directory structure:
+/// ```
+/// <cache_dir>/
+///   <mls_group_id_hex>/
+///     <subdirectory>/
+///       <filename>
+/// ```
+///
+/// This module handles:
+/// - Creating directory structures
+/// - Atomic file writes
+/// - File retrieval by path
+/// - File existence checks
+/// - Finding files by prefix
+///
+/// It does NOT handle:
+/// - Database operations (see database::media_files)
+/// - Network operations (see BlossomClient)
+/// - Encryption/decryption (caller's responsibility)
+pub struct MediaFileStorage {
+    cache_dir: PathBuf,
+}
+
+impl MediaFileStorage {
+    /// Creates a new MediaFileStorage instance
+    ///
+    /// # Arguments
+    /// * `data_dir` - The application data directory
+    ///
+    /// # Returns
+    /// A new MediaFileStorage instance with cache directory at `<data_dir>/media_cache/`
+    pub(crate) fn new(data_dir: &Path) -> Result<Self> {
+        let cache_dir = data_dir.join("media_cache");
+
+        // Create cache directory if it doesn't exist
+        if !cache_dir.exists() {
+            std::fs::create_dir_all(&cache_dir).map_err(|e| {
+                WhitenoiseError::MediaCache(format!("Failed to create cache directory: {}", e))
+            })?;
+        }
+
+        Ok(Self { cache_dir })
+    }
+
+    /// Stores a file in the cache using atomic write
+    ///
+    /// # Arguments
+    /// * `group_id` - The MLS group ID
+    /// * `subdirectory` - Subdirectory within the group (e.g., "group_images", "media")
+    /// * `filename` - The filename to store as (e.g., "abc123.jpg")
+    /// * `data` - The file data to store
+    ///
+    /// # Returns
+    /// The full path to the cached file
+    ///
+    /// # Errors
+    /// Returns error if filesystem operations fail
+    pub(crate) async fn store_file(
+        &self,
+        group_id: &GroupId,
+        subdirectory: &str,
+        filename: &str,
+        data: &[u8],
+    ) -> Result<PathBuf> {
```

Comment:
It could be useful for this method that it returns whether the file was already created or if it created it